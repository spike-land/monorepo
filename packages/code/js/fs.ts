import FS from "@isomorphic-git/lightning-fs";
// import FS from "@isomorphic-git/lightning-fs";

// import { fs } from 'memfs';

export const fs = new FS(location.origin);
const p = fs.promises;

// const readdir = globalThis.fs.readdir;

export const readdir = (filePath: string) => p.readdir(filePath);
export const writeFile = (filePath: string, content: string | Uint8Array) => p.writeFile(filePath, content);
export const readFile = (filePath: string) => p.readFile(filePath, { encoding: "utf8" });
export const stat = (filePath: string) => p.stat(filePath);
export const unlink = (filepath: string) => p.unlink(filepath);
export const mkdir = (filePath: string) => p.mkdir(filePath);
