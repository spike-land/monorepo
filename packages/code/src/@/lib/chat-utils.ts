import { replacePreservingWhitespace } from "@/lib/diff-utils";
import { Message } from "@/lib/interfaces";

const CODE_MODIFICATION_REGEX = /<<<<<<< SEARCH[\s\S]*?=======[\s\S]*?>>>>>>> REPLACE/g;
const MODIFICATION_SEPARATOR = ">>>>>>> REPLACE\n\n<<<<<<< SEARCH";
const SEARCH_REPLACE_MARKERS = ["<<<<<<< SEARCH", "=======", ">>>>>>> REPLACE"];

export const extractCodeModification = (response: string): string => {
  return response.match(CODE_MODIFICATION_REGEX)?.join("\n\n") || "";
};

export const loadMessages = (codeSpace: string): Message[] => {
  const key = `chatMessages-${codeSpace}`;
  const rawMessages = JSON.parse(localStorage.getItem(key) || "[]") as Message[];

  // Filter out messages without a role
  const validMessages = rawMessages.filter(m => !!m.role);

  // Remove consecutive messages with the same role
  const uniqueRoleMessages = validMessages.reduce((acc, current, index) => {
    if (index === 0 || current.role !== validMessages[index - 1].role) {
      acc.push(current);
    }
    return acc;
  }, [] as Message[]);

  return uniqueRoleMessages;
};

export const updateSearchReplace = (
  oldCode: string,
  codeNow: string,
  em = extractCodeModification,
): string => {
  const extractCodeModification = em;
  if (!oldCode.includes("<<<<<<< SEARCH")) {
    return codeNow;
  }

  try {
    const codeToReplace = extractCodeModification(oldCode);
    const modifications = codeToReplace
      .split(MODIFICATION_SEPARATOR)
      .filter((mod) => SEARCH_REPLACE_MARKERS.some((marker) => mod.includes(marker)))
      .map((mod) => mod.replace(/<<<<<<< SEARCH|>>>>>>> REPLACE/g, ""));

    return modifications.reduce((acc, modification) => {
      const [search, replace] = modification.split("=======");

      const result = replacePreservingWhitespace(
        acc,
        search.trim(),
        replace.split("\n").slice(1).map((x) => x.trim()).filter((x) => x).join(
          "\n",
        ),
      );

      console.table({
        success: result !== acc,
        search: search.trim(),
        replace,
        before: acc,
        after: result,
      });

      return result;
    }, codeNow);
  } catch (error) {
    console.error("Error during code modification:", error);
    return codeNow;
  }
};
