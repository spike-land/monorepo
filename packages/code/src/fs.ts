// import FS from "@isomorphic-git/lightning-fs";
// import FS from "@isomorphic-git/lightning-fs";
import { Mutex } from "async-mutex";
import * as memFS from "memfs";
import { importMapReplace } from "./importMapReplace";

// export const fs = new FS('fakeFS', {db: null});\

import FS from "@isomorphic-git/lightning-fs";

let fsProb: FS // | typeof memFS.fs;
;
try {
  if (typeof indexedDB === "undefined") fsProb = memFS.fs;
  else {
    // const FS = (await import("@isomorphic-git/lightning-fs")).default;
    fsProb = new FS("fakeFS");
  }
} catch {
  fsProb = memFS.fs;
}

export const fs = fsProb;

const p = fs.promises;

// const readdir = globalThis.fs.readdir;
const origin = typeof location !== "undefined" ? location.origin : "";

const mutex = new Mutex();

export const readdir = (filePath: string) =>
  mutex.runExclusive(() =>
    p.readdir(filePath).then((x) => x.map((d) => d.toString())).catch(() =>
      p.mkdir(filePath).then(() => p.readdir(filePath))
    )
  );
export const writeFile = (filePath: string, content: string | Uint8Array) =>
  mutex.runExclusive(() => {
    console.log("write", filePath);
    return p.writeFile(
      filePath,
      (filePath.indexOf(".js") !== -1 && filePath.indexOf(".json") === -1)
        ? `/*
written: ${new Date()}
  
*/
` + importMapReplace(content as string, origin, origin)
        : content,
    );
  });

export const readFile = (filePath: string) =>
  mutex.runExclusive(() =>
    p.readFile(filePath, { encoding: "utf8" }).catch(() => fetch(origin + filePath).then(x => x.text()))
  );
export const unlink = (filepath: string) =>
  mutex.runExclusive(() => p.unlink(filepath).catch(() => {/** nothing really happened */}));
export const mkdir = (filePath: string) => mutex.runExclusive(() => p.mkdir(filePath).catch(() => {}));

Object.assign(self, { readFile, unlink, mkdir, writeFile, readdir });
