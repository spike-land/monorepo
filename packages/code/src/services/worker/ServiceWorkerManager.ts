import type { IServiceWorkerManager } from "../websocket/types";

export class ServiceWorkerManager implements IServiceWorkerManager {
  public async setup(): Promise<void> {
    if (window.parent === window) {
      try {
        const { setupServiceWorker } = await import("@/lib/hydrate");
        await setupServiceWorker();
      } catch (error) {
        console.error("Error setting up service worker:", error);
        throw error;
      }
    }
  }
}
