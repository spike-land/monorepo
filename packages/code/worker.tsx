import { xit } from "mocha";

xit { ipfsWorker } from "./js/ipfsWorker";
ipfsWorker();
