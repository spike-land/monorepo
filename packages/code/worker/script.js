!function(e){var n={};function t(r){if(n[r])return n[r].exports;var o=n[r]={i:r,l:!1,exports:{}};return e[r].call(o.exports,o,o.exports,t),o.l=!0,o.exports}t.m=e,t.c=n,t.d=function(e,n,r){t.o(e,n)||Object.defineProperty(e,n,{enumerable:!0,get:r})},t.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},t.t=function(e,n){if(1&n&&(e=t(e)),8&n)return e;if(4&n&&"object"==typeof e&&e&&e.__esModule)return e;var r=Object.create(null);if(t.r(r),Object.defineProperty(r,"default",{enumerable:!0,value:e}),2&n&&"string"!=typeof e)for(var o in e)t.d(r,o,function(n){return e[n]}.bind(null,o));return r},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,"a",n),n},t.o=function(e,n){return Object.prototype.hasOwnProperty.call(e,n)},t.p="",t(t.s=0)}([function(e,n,t){"use strict";Object.defineProperty(n,"__esModule",{value:!0});const r=t(1);addEventListener("fetch",e=>{e.respondWith(r.handleRequest(e.request))})},function(e,n,t){"use strict";Object.defineProperty(n,"__esModule",{value:!0}),n.handleRequest=void 0;const r=t(2),o=SHATEST,i={"Access-Control-Allow-Origin":"*","Access-Control-Allow-Methods":"GET,HEAD,POST,OPTIONS","Access-Control-Max-Age":"86400"};n.handleRequest=async function(e){if("GET"===e.method)return e.url.endsWith("sw.js")?new Response(r.sw,{headers:{"content-type":"text/javascript"}}):new Response(r.html,{headers:{"content-type":"text/html"}});if("POST"===e.method){const n=await e.json(),t=(new TextEncoder).encode(JSON.stringify(n)),r=await crypto.subtle.digest({name:"SHA-256"},t),i=Array.from(new Uint8Array(r)).map(e=>("00"+e.toString(16)).slice(-2)).join("").substring(0,7);o.put(i,t);const s=new Response(`{"hash":"${i}"}`);return s.headers.append("Access-Control-Allow-Origin","*"),s.headers.append("Access-Control-Allow-Methods","GET,HEAD,POST,OPTIONS"),s.headers.append("Access-Control-Max-Age","86400"),s}return function(e){let n={...e.headers,...i,"Access-Control-Allow-Headers":e.headers.get("Access-Control-Request-Headers")};return new Response(e.body,{headers:n})}(e)}},function(e,n,t){"use strict";Object.defineProperty(n,"__esModule",{value:!0}),n.sw=n.html=n.version=void 0,n.version="7.0.23",n.html='<!DOCTYPE html>\n<html>\n\n<head>\n  <meta content="text/html;charset=utf-8" http-equiv="Content-Type">\n  <meta name="viewport" content="width=device-width,user-scalable=no,initial-scale=1">\n  <meta content="utf-8" http-equiv="encoding">\n\n  <style>\n    a,abbr,acronym,address,applet,article,aside,audio,b,big,blockquote,body,canvas,caption,center,cite,code,dd,del,details,dfn,div,dl,dt,em,embed,fieldset,figcaption,figure,footer,form,h1,h2,h3,h4,h5,h6,header,hgroup,html,i,iframe,img,ins,kbd,label,legend,li,mark,menu,nav,object,ol,output,p,pre,q,ruby,s,samp,section,small,span,strike,strong,sub,summary,sup,table,tbody,td,tfoot,th,thead,time,tr,tt,u,ul,var,video{margin:0;padding:0;border:0;font-size:100%;font:inherit;vertical-align:baseline}article,aside,details,figcaption,figure,footer,header,hgroup,menu,nav,section{display:block}body{line-height:1}ol,ul{list-style:none}blockquote,q{quotes:none}blockquote:after,blockquote:before,q:after,q:before{content:\'\';content:none}table{border-collapse:collapse;border-spacing:0}#container{background-color:#1e1e1e;width:100vw;height:100vh;animation-duration:1s;animation-name:opening}body{overflow:hidden;width:100%;height:100vh}\n\n    @keyframes opening {\n      from {\n        width: 10%;\n        height: 20vh;\n      }\n\n      66% {\n        width: 100%;\n        height: 20vh;\n      }\n\n      to {\n        width: 100%;\n        height: 100vh;\n      }\n    }\n\n\n    #error {\n      display: none;\n      background-color: red;\n      opacity: 0.7;\n    }\n\n    #root {\n      display: none;\n    }\n\n    #ace {\n      display: none;\n    }\n\n\n    .draggable {\n      margin: 24px;\n      padding: 32px;\n      position: absolute;\n      touch-action: none;\n      overflow: hidden;\n      z-index: 2;\n      word-wrap: break-word;\n      right: 24px;\n      /* float: right; */\n      /* top: 24px; */\n      /* right: 24px ; */\n      -webkit-transform: translate(0px, 0px);\n      transform: translate(0px, 0px);\n      font-size: 32px;\n      background-color: #ddd;\n      border-radius: 16px;\n      width: fit-content;\n      max-width: 40vw;\n      background: #ddd;\n      box-shadow:\n        0 2.8px 2.2px rgba(0, 0, 0, 0.034),\n        0 6.7px 5.3px rgba(0, 0, 0, 0.048),\n        0 12.5px 10px rgba(0, 0, 0, 0.06),\n        0 22.3px 17.9px rgba(0, 0, 0, 0.072),\n        0 41.8px 33.4px rgba(0, 0, 0, 0.086),\n        0 100px 80px rgba(0, 0, 0, 0.12);\n    }\n    #ace{\n      position: absolute;\n        top: 0;\n        right: 0;\n        bottom: 0;\n        left: 0;\n    }\n    \n\n    .almosthidden {\n      opacity: 0.5;\n    }\n\n    button {\n      font-size: large;\n    }\n  </style>\n</head>\n\n<body>\n  <div id="error" class="draggable"></div>\n  <div id="root" class="draggable resize-drag"></div>\n\n  <div id="container"></div>\n  <div id="ace"></div>\n  <script>\n  if (\'serviceWorker\' in navigator) {\n  navigator.serviceWorker.register(\'/sw.js\').then(function() {\n    console.log(\'service worker is is all cool.\');\n  }).catch(function(e) {\n    console.error(\'service worker is not so cool.\', e);\n    throw e;\n  });\n  }\n  // bling.js\n  var $ = window.$ = document.querySelector.bind(document);\n  var $$ = window.$$ = document.querySelectorAll.bind(document);\n  Node.prototype.on = window.on = function(name, fn) {\n    this.addEventListener(name, fn);\n  }\n  NodeList.prototype.__proto__ = Array.prototype;\n  NodeList.prototype.on = NodeList.prototype.addEventListener = (function(name, fn) {\n    this.forEach(function(elem) {\n      elem.on(name, fn);\n    });\n  });\n  <\/script>\n  \n  <script type="module">\n\n    const runner = async ()=>{\n      const version = "7.0.23";\n      const cdnAddress = "https://unpkg.com/@zedvision/code@" ;\n      const script = "/dist/_cBundle.js.min.js";\n\n\n      if (window.location.href.includes("0.0.0.0") || window.location.href.includes("localhost") ) {\n        const {run} = await import("./dist/_cBundle.js")\n        run();\n      }\n\n      const {run} = await import(cdnAddress + version + script)\n      run();\n    }\n    \n\n    //inject\n      \n    //inject\n\n    runner();\n  <\/script>\n</body>\n</html>',n.sw="var VERSION = '7';\n\nthis.addEventListener('install', function(e) {\n  e.waitUntil(caches.open(VERSION).then(cache => {\n    return cache.addAll([\n      '/',\n      '/index.html',\n      '/sw.js',\n    ]);\n  }))\n});\n\nthis.addEventListener('fetch', function(e) {\n  var tryInCachesFirst = caches.open(VERSION).then(cache => {\n    return cache.match(e.request).then(response => {\n      if (!response) {\n        return handleNoCacheMatch(e);\n      }\n      // Update cache record in the background\n      fetchFromNetworkAndCache(e);\n      // Reply with stale data\n      return response\n    });\n  });\n  e.respondWith(tryInCachesFirst);\n});\n\nthis.addEventListener('activate', function(e) {\n  e.waitUntil(caches.keys().then(keys => {\n    return Promise.all(keys.map(key => {\n      if (key !== VERSION)\n        return caches.delete(key);\n    }));\n  }));\n});\n\nfunction fetchFromNetworkAndCache(e) {\n  // DevTools opening will trigger these o-i-c requests, which this SW can't handle.\n  // There's probaly more going on here, but I'd rather just ignore this problem. :)\n  // https://github.com/paulirish/caltrainschedule.io/issues/49\n  if (e.request.cache === 'only-if-cached' && e.request.mode !== 'same-origin') return;\n\n  return fetch(e.request).then(res => {\n    // foreign requests may be res.type === 'opaque' and missing a url\n    if (!res.url) return res;\n    // regardless, we don't want to cache other origin's assets\n    if (new URL(res.url).origin !== location.origin) return res;\n\n    return caches.open(VERSION).then(cache => {\n      // TODO: figure out if the content is new and therefore the page needs a reload.\n      cache.put(e.request, res.clone());\n      return res;\n    });\n  }).catch(err => console.error(e.request.url, err));\n}\n\nfunction handleNoCacheMatch(e) {\n  return fetchFromNetworkAndCache(e);\n}"}]);