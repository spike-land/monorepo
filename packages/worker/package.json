{
  "private": false,
  "name": "@spike.land/spike-land-worker",
  "version": "0.0.4",
  "description": "A template for kick starting a Cloudflare Workers project",
  "main": "dist/index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "prebuild": "deno fmt src",
    "build": "rm -rf dist && mkdir dist && esbuild --bundle src/spike.land.handler.ts > dist/index.js",
    "deploy": "yarn build && wrangler publish"
  },
  "author": "Zoltan Erdos <zoltan.erdos@me.com>",
  "license": "MIT",
  "dependencies": {
    "@cloudflare/workers-types": "3.0.0",
    "@cloudflare/wrangler": "1.19.3",
    "@spike.land/code": "^0.0.4",
    "@spike.land/ipfs": "^0.0.4",
    "esbuild": "0.13.7",
    "ipfs-message-port-protocol": "0.10.1",
    "ipfs-message-port-server": "0.10.1",
    "ipfs-only-hash": "4.0.0",
    "uuid": "8.3.2"
  },
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "@types/uuid": "8.3.1"
  },
  "gitHead": "d0b137b49660b6106c2e1cab2a7f5d31345c4324"
}
