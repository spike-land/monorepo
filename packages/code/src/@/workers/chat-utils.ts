import { updateSearchReplace as up } from "@/lib/chat-utils";

export const updateSearchReplace = async (
  instructions: string,
  code: string,
) => {
  console.log("instructions", instructions);
  console.log("code", code);

  return up(instructions, code);
};

Object.assign(self, { updateSearchReplace });
