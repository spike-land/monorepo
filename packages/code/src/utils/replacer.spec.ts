import { updateSearchReplace } from "@/workers/chat-utils.worker";
import { readFile } from "node:fs/promises";
import { expect, test } from "vitest";

test("should update search replace", async () => {
  const original = await readFile(__dirname + "/landing-orig.txt", "utf8");
  const instructions = await readFile(__dirname + "/landing-inst.txt", "utf8");

  const r1 = await updateSearchReplace({ instructions, code: original });
  expect(r1.len).toMatchInlineSnapshot(`910`);

  const v2 = await updateSearchReplace({ instructions: instructions.slice(r1.len), code: r1.result });
  expect(v2.len).toMatchInlineSnapshot(`144`);

  const v3 = await updateSearchReplace({ instructions: instructions.slice(r1.len + v2.len), code: v2.result });
  expect(v3.len).toMatchInlineSnapshot(`170`);

  const v4 = await updateSearchReplace({ instructions: instructions.slice(r1.len + v2.len + v3.len), code: v3.result });
  expect(v4.len).toMatchInlineSnapshot(`4611`);

  const v5 = await updateSearchReplace({
    instructions: instructions.slice(r1.len + v2.len + v3.len + v4.len),
    code: v4.result,
  });
  expect(v5.len).toMatchInlineSnapshot(`411`);

  const v6 = await updateSearchReplace({
    instructions: instructions.slice(r1.len + v2.len + v3.len + v4.len + v5.len),
    code: v5.result,
  });
  expect(v6.len).toMatchInlineSnapshot(`732`);

  const v7 = await updateSearchReplace({
    instructions: instructions.slice(r1.len + v2.len + v3.len + v4.len + v5.len + v6.len),
    code: v6.result,
  });
  expect(v7.len).toMatchInlineSnapshot(`348`);

  const v8 = await updateSearchReplace({
    instructions: instructions.slice(r1.len + v2.len + v3.len + v4.len + v5.len + v6.len + v7.len),
    code: v7.result,
  });
  expect(v8.len).toMatchInlineSnapshot(`253`);

  const v9 = await updateSearchReplace({
    instructions: instructions.slice(r1.len + v2.len + v3.len + v4.len + v5.len + v6.len + v7.len + v8.len),
    code: v8.result,
  });
  expect(v9.len).toMatchInlineSnapshot(`342`);

  const v10 = await updateSearchReplace({
    instructions: instructions.slice(r1.len + v2.len + v3.len + v4.len + v5.len + v6.len + v7.len + v8.len + v9.len),
    code: v9.result,
  });
  expect(v10.len).toMatchInlineSnapshot(`219`);

  const v11 = await updateSearchReplace({
    instructions: instructions.slice(
      r1.len + v2.len + v3.len + v4.len + v5.len + v6.len + v7.len + v8.len + v9.len + v10.len,
    ),
    code: v10.result,
  });
  expect(v11.len).toMatchInlineSnapshot(`873`);

  const v12 = await updateSearchReplace({
    instructions: instructions.slice(
      r1.len + v2.len + v3.len + v4.len + v5.len + v6.len + v7.len + v8.len + v9.len + v10.len + v11.len,
    ),
    code: v11.result,
  });
  expect(v12.len).toMatchInlineSnapshot(`0`);

  expect(v12.result).toMatchSnapshot();
});

test("should update search replace with a single block", async () => {
  const original = await readFile(__dirname + "/aclock-orig.txt", "utf8");
  const instructions = await readFile(__dirname + "/acllock-inst.txt", "utf8");

  const r1 = await updateSearchReplace({ instructions, code: original });
  expect(r1.len).toMatchInlineSnapshot(`2358`);

  const v2 = await updateSearchReplace({ instructions: instructions.slice(r1.len), code: r1.result });
  expect(v2.len).toMatchInlineSnapshot(`541`);

  const v3 = await updateSearchReplace({ instructions: instructions.slice(r1.len + v2.len), code: v2.result });
  expect(v3.len).toMatchInlineSnapshot(`454`);

  const v4 = await updateSearchReplace({ instructions: instructions.slice(r1.len + v2.len + v3.len), code: v3.result });
  expect(v4.len).toMatchInlineSnapshot(`1046`);

  const v5 = await updateSearchReplace({
    instructions: instructions.slice(r1.len + v2.len + v3.len + v4.len),
    code: v4.result,
  });
  expect(v5.len).toMatchInlineSnapshot(`0`);

  expect(v5.result).toMatchSnapshot();
});
