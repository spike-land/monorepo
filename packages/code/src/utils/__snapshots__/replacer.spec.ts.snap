// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should update search replace 13`] = `
  "// landing.tsx
  import React, { useState, useEffect, createContext, useContext } from "react";
  import { ImageLoader } from "@/components/ui/image-loader";
  import { Card, CardContent } from "@/components/ui/card";
  import { Code, Zap, Bot, Share2, Package, Download } from "lucide-react";
  import { motion } from "framer-motion";
  import Prompt from "@/components/ui/start-with-prompt";

  //bubu
  // Creathey hellope
  const ThemeContext = createContext({
    isDarkMode: false,
    toggleTheme: () => {},
  });

  // Feature component
  const Feature = ({ icon: Icon, title, description, color, imagePrompt }) => {
    return (
      <motion.div
        initial={{ opacity: 0, y: 20 }}
        animate={{ opacity: 1, y: 0 }}
        transition={{ duration: 0.5 }}
        whileHover={{ scale: 1.03 }}
        whileInView={{
          y: [0, -10, 0],
          transition: {
            y: {
              repeat: Infinity,
              repeatType: "reverse",
              duration: 2,
              ease: "easeInOut",
            },
          },
        }}>
        <Card className='h-full hover:shadow-2xl transition-all duration-500 bg-white/30 dark:bg-gray-800/50 backdrop-blur-lg border border-white/30 rounded-xl overflow-hidden transform hover:scale-105'>
          <CardContent className='p-4 flex flex-col items-center text-center group'>
            <div className='mb-2 w-full h-24 overflow-hidden rounded-lg'>
              <ImageLoader
                prompt={imagePrompt}
                className='w-full h-full object-cover transition-all duration-300 transform group-hover:scale-110'
              />
            </div>
            <div className='icon-wrapper mb-2'>
              <Icon
                className={\`w-8 h-8 \${color} group-hover:scale-110 transition-all duration-300 transform group-hover:rotate-12\`}
              />
            </div>
            <h3 className='text-lg sm:text-xl font-bold mb-2 text-white'>{title}</h3>
            <p className='text-sm sm:text-base text-white/90'>{description}</p>
          </CardContent>
        </Card>
      </motion.div>
    );
  };

  const LandingPage = () => {
    const [isDarkMode, setIsDarkMode] = useState(false);

    const toggleTheme = () => {
      setIsDarkMode((prevMode) => {
        const newMode = !prevMode;
        localStorage.setItem("theme", newMode ? "dark" : "light");
        return newMode;
      });
    };

    useEffect(() => {
      const savedTheme = localStorage.getItem("theme");
      if (savedTheme === "dark") {
        setIsDarkMode(true);
      }
    }, []);

    const features = [
    {
      icon: Bot,
      title: "Intelligens Virtuális Asszisztensek",
      description:
        "Forradalmasítsa az ügyfélkapcsolatokat MI-vezérelt virtuális asszisztenseinkkel, amelyek képesek összetett kérdések kezelésére és minden interakcióból tanulnak.",
      color: "text-red-600",
      imagePrompt:
        "Egy futurisztikus holografikus MI-asszisztens kommunikál az ügyfelekkel",
    },
    {
      icon: Zap,
      title: "Fejlett Prediktív Analitika",
      description:
        "Használja ki a big data erejét élvonalbeli prediktív modelljeinkkel, amelyek lehetővé teszik a piaci trendek előrejelzését és az informált döntéshozatalt.",
      color: "text-white",
      imagePrompt:
        "3D holografikus kijelző, amely összetett prediktív modelleket és adatáramlásokat mutat",
    },
    {
      icon: Code,
      title: "Egyedi MI-fejlesztés",
      description:
        "MI-szakértőink csapata testreszabott megoldásokat hoz létre az Ön egyedi üzleti kihívásainak kezelésére és az innováció elősegítésére.",
      color: "text-green-600",
      imagePrompt:
        "MI-fejlesztők együttműködése egy high-tech, futurisztikus irodatérben",
    },
    {
      icon: Share2,
      title: "Zökkenőmentes MI-integráció",
      description:
        "Erőfeszítés nélkül építse be az MI-képességeket meglévő infrastruktúrájába fejlett integrációs keretrendszereinkkel és API-jainkkal.",
      color: "text-red-600",
      imagePrompt:
        "A zökkenőmentes MI-integráció vizuális ábrázolása különböző üzleti rendszerekkel",
    },
    {
      icon: Package,
      title: "Kognitív Folyamatautomatizálás",
      description:
        "Alakítsa át működését MI-vezérelt automatizálási megoldásainkkal, amelyeket összetett feladatok kezelésére és a szervezet egészében a hatékonyság javítására terveztünk.",
      color: "text-white",
      imagePrompt:
        "MI-vezérelt robotok és digitális rendszerek harmonikus együttműködése egy futurisztikus gyárban",
    },
    {
      icon: Download,
      title: "Adaptív MI-ökoszisztémák",
      description:
        "Maradjon az innováció élvonalában önfejlesztő MI-modelljeinkkel, amelyek folyamatosan alkalmazkodnak az új adatokhoz és a változó üzleti környezethez.",
      color: "text-green-600",
      imagePrompt:
        "Egy dinamikus, fejlődő MI-ökoszisztéma ábrázolása összekapcsolt neurális hálózatokként",
    },
  ];

    return (
      <ThemeContext.Provider value={{ isDarkMode, toggleTheme }}>
        <div className='min-h-screen flex flex-col bg-gradient-to-br from-red-500 via-white to-green-500 text-white relative'>
          <div
            className='absolute inset-0 opacity-10 bg-repeat'
            style={{
              backgroundImage:
                "url('data:image/svg+xml,%3Csvg width='60' height='60' viewBox='0 0 60 60' xmlns='http://www.w3.org/2000/svg'%3E%3Cg fill='none' fill-rule='evenodd'%3E%3Cg fill='%23ffffff' fill-opacity='1'%3E%3Cpath d='M36 34v-4h-2v4h-4v2h4v4h2v-4h4v-2h-4zm0-30V0h-2v4h-4v2h4v4h2V6h4V4h-4zM6 34v-4H4v4H0v2h4v4h2v-4h4v-2H6zM6 4V0H4v4H0v2h4v4h2V6h4V4H6z'/%3E%3C/g%3E%3C/g%3E%3C/svg%3E')",
            }}></div>

          <div className='absolute top-0 left-0 w-full h-full pointer-events-none'>
            <svg
              className='w-full h-full opacity-5'
              xmlns='http://www.w3.org/2000/svg'
              viewBox='0 0 100 100'>
              <path
                d='M30,10 L50,50 L70,10 M30,90 L50,50 L70,90'
                fill='none'
                stroke='white'
                strokeWidth='0.5'
              />
              <circle
                cx='50'
                cy='50'
                r='40'
                fill='none'
                stroke='white'
                strokeWidth='0.5'
              />
            </svg>
          </div>
          <header className='py-4 bg-red-600/10 backdrop-blur-lg text-white shadow-lg'>
            <div className='container mx-auto text-center'>
              <h1 className='text-2xl sm:text-3xl md:text-4xl font-bold mb-2'>
    Üdvözöljük a NeuralNexnél
  </h1>
  <marquee className='text-base sm:text-lg mb-4'>
    Szabadítsa fel a mesterséges intelligencia erejét a NeuralNexszel!
  </marquee>
  <p className='text-xl mb-4'>
    Emelje vállalkozását a következő szintre élvonalbeli MI-megoldásainkkal
  </p>
              <button
                onClick={() => {
                  window.location.href = "/start";
                }}
                className='bg-white/20 hover:bg-white/30 backdrop-blur-lg text-white text-base font-bold py-2 px-6 rounded-full transition-all duration-300 border border-white/30'>
                MI-lehetőségek felfedezése
              </button>
            </div>
          </header>
          <div className='h-1 bg-white/20 backdrop-blur-lg'></div>
  <main className='flex-grow bg-white/10 backdrop-blur-lg py-24 relative overflow-hidden'>
            <div className='container mx-auto px-4'>
              <h2 className='text-3xl sm:text-4xl font-bold mb-8 sm:mb-12 text-center text-white'>
                Átalakító MI-megoldások minden iparág számára
              </h2>
              <div className='grid grid-cols-1 md:grid-cols-2 xl:grid-cols-3 gap-4 sm:gap-6 lg:gap-8'>
                {features.map((feature, index) => (
                  <Feature
                    key={index}
                    {...feature}
                    imagePrompt={feature.imagePrompt}
                  />
                ))}
              </div>
            </div>
          </main>

          <footer className='bg-green-600/10 backdrop-blur-lg py-8 sm:py-12 text-white shadow-lg'>
            <div className='container mx-auto px-4'>
              <div className='flex flex-col sm:flex-row justify-between items-center'>
                <div className='mb-6 sm:mb-0'>
                  <h3 className='text-xl sm:text-2xl font-bold mb-2'>NeuralNex</h3>
  <p className='text-sm sm:text-base text-gray-400'>
    &copy; 2024 NeuralNex. Minden jog fenntartva.
  </p>
                </div>
                <nav className='flex flex-col sm:flex-row space-y-4 sm:space-y-0 sm:space-x-6'>
                  <a
    href='/solutions'
    className='hover:text-white transition-colors duration-300'>
    MI-megoldások
  </a>
  <a
    href='/case-studies'
    className='hover:text-red-400 transition-colors duration-300'>
    Sikertörténetek
  </a>
  <a
    href='/contact'
    className='hover:text-red-400 transition-colors duration-300'>
    Partnerség
  </a>
                </nav>
              </div>
            </div>
          </footer>
        </div>
      </ThemeContext.Provider>
    );
  };

  export default LandingPage;

//1726708s
"
`;
